rumqttc/src/tls.rs

```diff
@@ -106,27 +107,19 @@ pub async fn rustls_connector(tls_config: &TlsConfiguration) -> Result<RustlsCon
                     rustls_pemfile::certs(&mut BufReader::new(Cursor::new(client.0.clone())))?;
                 // load appropriate Key as per the user request. The underlying signature algorithm
                 // of key generation determines the Signature Algorithm during the TLS Handskahe.
-                let read_keys = match &client.1 {
-                    Key::RSA(k) => rustls_pemfile::rsa_private_keys(&mut BufReader::new(
-                        Cursor::new(k.clone()),
-                    )),
-                    Key::ECC(k) => rustls_pemfile::pkcs8_private_keys(&mut BufReader::new(
-                        Cursor::new(k.clone()),
-                    )),
-                };
-                let keys = match read_keys {
-                    Ok(v) => v,
-                    Err(_e) => return Err(Error::NoValidCertInChain),
-                };
 
-                // Get the first key. Error if it's not valid
-                let key = match keys.first() {
-                    Some(k) => k.clone(),
-                    None => return Err(Error::NoValidCertInChain),
+                // Create buffer for key file
+                let mut key_buffer = BufReader::new(Cursor::new(client.1.clone()));
+                // We only read the first key in the key file
+                let key = match rustls_pemfile::read_one(&mut key_buffer) {
+                    Ok(Some(rustls_pemfile::Item::RSAKey(key)))
+                    | Ok(Some(rustls_pemfile::Item::PKCS8Key(key)))
+                    | Ok(Some(rustls_pemfile::Item::ECKey(key))) => key,
+                    Ok(None) | Ok(Some(_)) => return Err(Error::NoValidKey),
+                    Err(err) => return Err(Error::Io(err)),
```

Comment:
in case if it is `Some(_)` but not key, we want to continue looking for the key ( because that is how fns like `rsa_private_keys` work!, they ignore the other items ).

```rust
// keep reading Items one by one to find a Key, return error if none found.
let key = loop {
    // note: as read_one returns io::Error, and as our Error::IO(_) has from io::Error
    // i just put ? here. if this is not working, we might have to use .map_err()?
    let item = rustls_pemfile::read_one(key_buffer)?;
    match item {
        Some(Item::ECKey(key) | Item::RSAKey(key) | Item::PKCS8Key(key)) => {
            break key;
        }
        // added that InChain cuz we have it for Certs lol
        None => return Err(Error::NoValidKeyInChain),
        _ => {}
    }
}
```

Reply:
Ah good point!

Reply:
Done